### PR TITLE
Back to Devise

### LINKED ISSUE #68: Password login with default password for drivers
Jamin pointed out that while there's no real incentive for drivers to log in as each other, badge number + last name isn't secure enough for staff, who have access to sensitive data.

Proposed implementing a system where staff would have to change their passwords, which would by default be their last name.

Should run by the group before implementing.

### PR DESCRIPTION
**Pull request by dfaulken**
Closes #68.

Should follow this up with `RAILS_ENV=production bundle exec rake passwords:set_all_default`.

New login screen:
screen shot 2017-12-20 at 2 27 15 pm

Supervisors and staff must change their passwords:
screen shot 2017-12-20 at 2 27 22 pm
screen shot 2017-12-20 at 2 27 52 pm

Staff can reset people's passwords to the default:
screen shot 2017-12-20 at 2 28 05 pm
screen shot 2017-12-20 at 2 28 15 pm

Had to change `@current_user` to `current_user` everywhere, so sorry for the noise.

### CODE REVIEW

File: config/initializers/devise.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Devise.setup do |config|
+  require 'devise/orm/active_record'
+
+  config.mailer_sender = 'incidents@pvtaapps.com'
+  config.authentication_keys = [:badge_number]
+  config.case_insensitive_keys = [:badge_number]
+  config.strip_whitespace_keys = [:badge_number]
+  config.stretches = Rails.env.test? ? 1 : 11
```
Comment (dfaulken) [inline config/initializers/devise.rb]: Number of times the encryption algorithm is applied to the password.

File: app/models/user.rb [hunk 2]
```diff
@@ -43,10 +51,30 @@ def group
...
+    else super
+    end
+  end
+
   def proper_name
     [last_name, first_name].join ', '
   end
 
+  def requires_password_change?
+    !(driver? || password_changed_from_default?)
```
Comment (dfaulken) [inline app/models/user.rb]: That is, drivers and people who have changed their password, do not need to change their password.

File: config/routes.rb [hunk 3]
```diff
@@ -2,6 +2,13 @@
 
 Rails.application.routes.draw do
   root 'incidents#index'
+
+  devise_for :users, skip: :registrations
+  devise_scope :user do
+    get 'users/change_password' => 'devise/registrations#edit', as: :change_password
+    patch 'users/:id' => 'devise/registrations#update', as: :user_registration
```
Comment (dfaulken) [inline config/routes.rb]: I still sort of despise this notation

File: db/migrate/20171219190734_add_devise_to_users.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+      t.string :encrypted_password, null: false, default: ""
+      t.string   :reset_password_token
+      t.datetime :reset_password_sent_at
+    end
+
+    add_index :users, :reset_password_token, unique: true
+  end
+
+  def self.down
+    raise ActiveRecord::IrreversibleMigration
```
Comment (Anbranin) [inline db/migrate/20171219190734_add_devise_to_users.rb]: ooooh

File: app/models/user.rb (hunk 2 above)
Comment (Anbranin) [reply app/models/user.rb]: How do you tell? Is this a devise thing?

File: app/models/user.rb [hunk 5]
```diff
@@ -92,4 +120,12 @@ def self.import_from_xml(xml)
     statuses[:deactivated] = inactive_users.count
     statuses
   end
+
+  private
+
+  def track_password_changed
+    if encrypted_password_changed?
+      assign_attributes password_changed_from_default: true
```
Comment (Anbranin) [inline app/models/user.rb]: Oh now I get it

File: db/migrate/20171219190734_add_devise_to_users.rb (hunk 4 above)
Comment (dfaulken) [reply db/migrate/20171219190734_add_devise_to_users.rb]: Yeah, this comes from Devise.

File: app/models/user.rb (hunk 2 above)
Comment (dfaulken) [reply app/models/user.rb]: https://github.com/umts/incidents/pull/89#discussion_r158549973

**Pull request merged**